### PR TITLE
fix: blocknumber binsearch

### PR DESCRIPTION
**Pull request by 0xyaco**
# 🤖 Linear

Closes GRT-226

## Description
* Handle the cases where multiple blocks have the same timestamp (in arbitrum multiple blocks can have the same timestamp if working with the precision in seconds)
* Fixed precision errors when the estimated block time is between [0, 1)

### CODE REVIEW

File: packages/blocknumber/package.json [hunk 1]
```diff
@@ -25,6 +25,7 @@
     "dependencies": {
         "@ebo-agent/shared": "workspace:*",
         "axios": "1.7.7",
+        "bignumber.js": "^9.1.2",
```
[the remark this answers is not in the document]
Comment (0xyaco) [reply packages/blocknumber/package.json:28]: NOOO ITS BACK  :axe:

File: packages/blocknumber/src/providers/evmBlockNumberProvider.ts [hunk 2]
```diff
@@ -234,4 +255,55 @@ export class EvmBlockNumberProvider implements BlockNumberProvider {
...
+        }
+    }
+
+    /**
+     * Search the block with the lowest height that has the same timestamp as `block`.
+     *
+     * @param block the block to use in the search
+     * @returns a block with the same timestamp as `block` and with the lowest height.
+     */
+    private async searchFirstBlockWithEqualTimestamp(
```
Comment (0xnigir1) [inline packages/blocknumber/src/providers/evmBlockNumberProvider.ts:294]: maybe call this `getFirstBlockWithEqualTimestamp` or the other function `searchNextBlockWithDifferentTimestamp`, so you keep a naming convention for search/get functions

File: packages/blocknumber/src/providers/evmBlockNumberProvider.ts [hunk 3]
```diff
@@ -129,17 +130,24 @@ export class EvmBlockNumberProvider implements BlockNumberProvider {
...
-        const estimatedBlockTime = await this.estimateBlockTime(lastBlock, blocksLookback);
-        const timestampDelta = lastBlock.timestamp - timestamp;
-        let candidateBlockNumber = lastBlock.number - timestampDelta / estimatedBlockTime;
+        const estimatedBlockTimeBN = await this.estimateBlockTime(lastBlock, blocksLookback);
+        const timestampDeltaBN = new BigNumber((lastBlock.timestamp - timestamp).toString());
 
-        const baseStep = (lastBlock.number - candidateBlockNumber) * deltaMultiplier;
+        let candidateBlockNumberBN = new BigNumber(lastBlock.number.toString())
+            .dividedBy(timestampDeltaBN.dividedBy(estimatedBlockTimeBN))
+            .integerValue();
```
Comment (0xnigir1) [inline packages/blocknumber/src/providers/evmBlockNumberProvider.ts:138]: is this okay? original code is: `let candidateBlockNumber = lastBlock.number - timestampDelta / estimatedBlockTime;` and you're doing two divisions (idk if im missing smth on the math)

**Review by 0xnigir1 — COMMENTED** [review]
just one doubt, else lgtm

File: packages/blocknumber/src/providers/evmBlockNumberProvider.ts (hunk 2 above)
Comment (0xyaco) [reply packages/blocknumber/src/providers/evmBlockNumberProvider.ts]: Nice suggestion, I think I'll go with `search`; feels like a more appropriate naming for `O(n)` lookups. 

`get` feels like more appropriate for `O(1)` functions, wdty?

File: packages/blocknumber/src/providers/evmBlockNumberProvider.ts (hunk 3 above)
Comment (0xyaco) [reply packages/blocknumber/src/providers/evmBlockNumberProvider.ts]: Oh boy what a nice catch :dart:

**Review by 0xnigir1 — APPROVED** [review]

**Pull request merged**